### DISCUSSION TITLE
Using edge attr

### DISCUSSION THREAD

**Discussion by JindeShubhamA** [pyg-team/pytorch_geometric#2495] [Q&A]
Hello,

I have spent around 2 months on this repository and done some benchmarking on synthetic datasets for both Graph and Node classification tasks using GCN, SageConv, and Attention-based Graph networks. I would say this repository is the best that is out there for performing experiments on Graph-based networks. 

The dataset I worked on till now did not have edge attributes. However know as I have become somewhat comfortable, I would like to also experiment with the edge attribute task. Edge attributes will be the node features that are passed during message propagation. 
1) Can you direct on some examples where edge attributes are used?
2) Also are there some GNN layers that use edge attributes in their default implementation?

**Comment by rusty1s** [comment]
Thanks for your kind words :)

Not all GNN operators support edge features. This is mostly due to following the precise formulation of a given paper. While on can naturally incorporate edge features in the message passing phase, there exist multiple ways to do so (e.g. via summation or concatenation). The best way to find all GNN operators that can make use of edge features is to search for `edge_attr` in the `torch_geometric.nn` documentation. Here are a few out-of-my-head: `GENConv`, `GINEConv`, `GMMConv`, `TransformerConv`, `PNAConv`, ...

[Here](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/pna.py) is an example that makes use of edge features.

**Reply by JindeShubhamA** [reply]
Thanks, I can explore the GNN operators mentioned above

I was thinking will it be possible somehow to use  **`def message(self, x_j, edge_attr):`** function with GNN operators such as GCNConv  where you can append node features with edge features?

In a very simple example if there is a directed connection from node B to node A having some **`edge_attr,`** can we not append this **`edge_attr`** to the features of B which is passed to A?

**Reply by rusty1s** [reply]
Yes, that is the standard approach, i.e., `self.lin(torch.cat([x_j, edge_attr], dim=-1))`. A different approach is summation, i.e., `x_j + self.lin(edge_attr)`.

**Reply by JindeShubhamA** [reply]
okay yes I also thought the same. Can you maybe suggest me on possible ways of implementing this. Do I need to change the current implementation of GCNConv with different message function that also has **`edge_attr`** as a parameter. I suspect for that  I will need to create **`gcn_conv_new.py`** file which has the GCNConv class, all the helper functions from the original **`gcn_conv.py`**  and updated **`message`** function.

Is there any smatter way to do it to avoid redundant code?

**Reply by rusty1s** [reply]
Currently yes, you need to create a fully new-version of it. It might be also possible to implement something like this by making use of internal state, i.e.:
```python
class MyGCNConv(GCNConv):
    def _init__(self, *args, edge_dim=0, **kwargs):
        super().__init__(*args, **kwargs):
        self.edge_lin = Linear(edge_dim, self.out_channels)
        
    def forward(self, x, edge_index, edge_attr):
        self._edge_attr = edge_attr
        out = super().forward(x, edge_index)
        self._edge_attr = None
        return out

    def message(self, x_j, edge_weight):
        return (x_j + self.edge_lin(self._edge_attr))) * edge_weight.view(-1, 1)
```
I need to think of better ways to handle such customization though.

**Reply by JindeShubhamA** [reply]
Found a way around. It is getting trained. Only issue is GCNConv adds self loop which increases number of edges and edge_attr does not have any values for it. Currently training by removing the self loop which does not make much sense. But yes I can tackle this through multiple ways. Thanks

**Reply by a-ism256** [reply]
Hello,

Do you mind sharing what ended up working for you?

I'm trying to do the same thing for SAGEConv and this thread seems to be the closest to verifying a working solution.

According to what @rusty1s said above, I'm implementing as follows. Also, I understand that this is the summation approach not the concatenation. Not sure if concatenation would be worth trying.

```
class EdgeSAGEConv(SAGEConv):
    def _init__(self, *args, edge_dim=0, **kwargs):
        super().__init__(*args, **kwargs)
        self.edge_lin = Linear(edge_dim, self.out_channels)
        
    def forward(self, x, edge_index, edge_attr):
        self._edge_attr = edge_attr
        out = super().forward(x, edge_index)
        self._edge_attr = None
        return out

    def message(self, x_j, edge_weight):
        return (x_j + self.edge_lin(self._edge_attr)) * edge_weight.view(-1, 1)
```

Then, the rest of the model initialization and training would be as in the [reddit](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/reddit.py) example. I haven't tested it yet, just trying to understand everything and make sure it all makes sense before I do.

Thanks!

**Reply by rusty1s** [reply]
Yes, this should work :)

**Reply by JindeShubhamA** [reply]
Hello,
I did try the method you are trying but had few issues. So I had a new version of GCNConv in which I am taking care of edge_attr as node features. As I wanted the same for SAGEConv, GATConv and TAGConv, I ended up spending sometime looking at the implementation of this standard layers and just modified them.
